### PR TITLE
[damtomo] add support for Damtomo service

### PR DESCRIPTION
**Pull request by nao20010128nao**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Damtomo is a service to expose users' singing scenes.

example video url: https://www.clubdam.com/app/damtomo/karaokeMovie/StreamingDkm.do?karaokeMovieId=2414316
example record url: https://www.clubdam.com/app/damtomo/karaokePost/StreamingKrk.do?karaokeContributeId=27376862

### CODE REVIEW

File: yt_dlp/extractor/damtomo.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+            raise ExtractorError('There is an error on server-side. Try again later.', expected=True)
+
+        # NOTE: there are excessive amount of spaces and line breaks, so ignore spaces around these part
+        description = self._search_regex(r'(?m)<div id="public_comment">\s*<p>\s*([^<]*?)\s*</p>', webpage, 'description', default=None)
+        uploader_id = self._search_regex(r'<a href="https://www\.clubdam\.com/app/damtomo/member/info/Profile\.do\?damtomoId=([^"]+)"', webpage, 'uploader_id', default=None)
+
+        # cleaner way to extract information in HTML
+        # example content: https://gist.github.com/nao20010128nao/1d419cc9ca3177be134094addf28ab51
+        data_dict = {g.group(2): clean_html(g.group(3)) for g in re.finditer(r'(?s)<(p|div)\s+class="([^" ]+?)">(.+?)</\1>', webpage)}
+        data_dict = {k: re.sub(r'\s+', ' ', v) for k, v in data_dict.items() if v}
```
Comment (pukkandan) [inline yt_dlp/extractor/damtomo.py:28]:
```suggestion
        data_dict = {
            mobj.group('class'): re.sub(r'\s+', ' ', clean_html(mobj.group('value')))
            for mobj in re.finditer(r'(?s)<(p|div)\s+class="(?P<class>[^" ]+?)">(?P<value>.+?)</\1>', webpage)}
```

File: yt_dlp/extractor/damtomo.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+        webpage, handle = self._download_webpage_handle(self._WEBPAGE_URL_TMPL % video_id, video_id, encoding='sjis')
+
+        if handle.url == 'https://www.clubdam.com/sorry/':
+            raise ExtractorError('You are rate-limited. Try again later.', expected=True)
+        if '<h2>予期せぬエラーが発生しました。</h2>' in webpage:
+            raise ExtractorError('There is an error on server-side. Try again later.', expected=True)
+
+        # NOTE: there are excessive amount of spaces and line breaks, so ignore spaces around these part
```
Comment (pukkandan) [inline yt_dlp/extractor/damtomo.py:21]:
```suggestion
```
using /s* everywhere is already recommended practice

File: yt_dlp/extractor/damtomo.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+            'uploader_id': uploader_id,
+            'description': description,
+            'uploader': data_dict['user_name'],
+            'upload_date': try_get(data_dict, lambda x: self._search_regex(r'(\d\d\d\d/\d\d/\d\d)', x['date'], 'upload_date', default=None).replace('/', ''), compat_str),
+            'view_count': int_or_none(self._search_regex(r'(\d+)', data_dict['audience'], 'view_count', default=None)),
+            'like_count': int_or_none(self._search_regex(r'(\d+)', data_dict['nice'], 'like_count', default=None)),
+            'track': title,
+            'artist': data_dict['song_artist'],
+
+            'url': m3u8_url,
```
Comment (pukkandan) [inline yt_dlp/extractor/damtomo.py:56]: dont pass m3u8 url directly. parse and sort the formats

File: yt_dlp/extractor/damtomo.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+            'id': video_id,
+            'title': title,
+            'uploader_id': uploader_id,
+            'description': description,
+            'uploader': data_dict['user_name'],
+            'upload_date': try_get(data_dict, lambda x: self._search_regex(r'(\d\d\d\d/\d\d/\d\d)', x['date'], 'upload_date', default=None).replace('/', ''), compat_str),
+            'view_count': int_or_none(self._search_regex(r'(\d+)', data_dict['audience'], 'view_count', default=None)),
+            'like_count': int_or_none(self._search_regex(r'(\d+)', data_dict['nice'], 'like_count', default=None)),
+            'track': title,
+            'artist': data_dict['song_artist'],
```
Comment (pukkandan) [inline yt_dlp/extractor/damtomo.py:54]: .get

File: yt_dlp/extractor/damtomo.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+            './/d:streamingUrl', {'d': self._DKML_XML_NS}).text.strip(), compat_str)
+        if not m3u8_url:
+            raise ExtractorError('Failed to obtain m3u8 URL')
+
+        return {
+            'id': video_id,
+            'title': title,
+            'uploader_id': uploader_id,
+            'description': description,
+            'uploader': data_dict['user_name'],
```
Comment (pukkandan) [inline yt_dlp/extractor/damtomo.py:49]: .get

File: yt_dlp/extractor/damtomo.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+            'upload_date': try_get(data_dict, lambda x: self._search_regex(r'(\d\d\d\d/\d\d/\d\d)', x['date'], 'upload_date', default=None).replace('/', ''), compat_str),
```
Comment (pukkandan) [inline yt_dlp/extractor/damtomo.py:50]: use unified_strdate

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/damtomo.py [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+            m3u8_url, video_id, ext='mp4')
+        self._sort_formats(formats)
+
+        return {
+            'id': video_id,
+            'title': title,
+            'uploader_id': uploader_id,
+            'description': description,
+            'uploader': data_dict.get('user_name'),
+            'upload_date': unified_strdate(self._search_regex(r'(\d\d\d\d/\d\d/\d\d)', data_dict.get('date'), 'upload_date', default=None)),
```
Comment (pukkandan) [inline yt_dlp/extractor/damtomo.py:51]:
```suggestion
            'upload_date': unified_strdate(self._search_regex(r'(\d{4}/\d{2}/\d{2})', data_dict.get('date'), 'upload_date', default=None)),
```

File: yt_dlp/extractor/damtomo.py [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+            self._DKML_XML_URL % video_id, video_id, note='Requesting stream information', encoding='sjis',
+            # doing this has no problem since there is no character outside ASCII,
+            # and never likely to happen in the future
+            transform_source=lambda x: re.sub(r'\s*encoding="[^"]+?"', '', x))
+        m3u8_url = try_get(stream_tree, lambda x: x.find(
+            './/d:streamingUrl', {'d': self._DKML_XML_NS}).text.strip(), compat_str)
+        if not m3u8_url:
+            raise ExtractorError('Failed to obtain m3u8 URL')
+        formats = self._extract_m3u8_formats(
+            m3u8_url, video_id, ext='mp4')
```
Comment (pukkandan) [inline yt_dlp/extractor/damtomo.py:42]:
```suggestion
        formats = self._extract_m3u8_formats(m3u8_url, video_id, ext='mp4')
```

**Review by pukkandan — APPROVED** [review]
just a few tiny changes. then it can be merged

File: yt_dlp/extractor/damtomo.py (hunk 8 above)
Comment (nao20010128nao) [reply yt_dlp/extractor/damtomo.py]: All of them done

**Pull request merged**